### PR TITLE
UID2-4391 Major refactor to improve performance and maintainability

### PR DESCRIPTION
**Pull request by atarassov-ttd**
- Removed redundant separation of IO and Worker threads keeping a single pool IO worker threads
    - Removed memory arena and custom memory management for IO buffers
    - Removed half-baked support for multiple inflight output buffers
    - Increased IO buffer size to 4096 bytes
    - Simplified and cleaned up a lot of code
    - Removed support for socks proxy in config as this was not an implemented feature
    - Simplified socket implementation, especially around how sockets get closed in case of errors
    - Make proxy a bit fairer by round-robining across all connections with a single IO operation at a time, instead of draining each socket in order
    - Make support for handling async connections more explicit
    - Add unit test coverage, most importantly - for the sockets IO logic
    - Improved/corrected README

    Manually verified through various tests as well as using tcptunnelchecker.

### CODE REVIEW

File: vsock-bridge/src/config.cpp [hunk 1]
```diff
@@ -36,41 +31,38 @@ namespace vsockproxy
...
-		case endpoint_scheme::VSOCK: return "vsock";
+		case EndpointScheme::TCP4: return "tcp";
+		case EndpointScheme::VSOCK: return "vsock";
 		default: return "unknown";
 		}
 	}
 
 
-	uint16_t try_str2short(std::string s, uint16_t default_value)
+    static std::optional<uint16_t> tryStr2Short(const std::string& s)
```
Comment (scong-ttd) [inline vsock-bridge/src/config.cpp]: nit: the algorithm is actually "str to ushort". And we probably should just call `stoi`.

**Review by thomasm-ttd — APPROVED** [review]

File: vsock-bridge/src/socket.cpp [hunk 2]
```diff
@@ -17,142 +17,86 @@ namespace vsockio
...
-				// Error
-
-				Logger::instance->Log(Logger::WARNING, "[socket] error on read, closing input (fd=", _fd, "): ", strerror(err));
-				closeInput();
-				break;
-			}
-		}
-
-		return buffer;
+        if (!buffer.hasRemainingCapacity()) return false;
```
Comment (scong-ttd) [inline vsock-bridge/src/socket.cpp:56]: indentation?

File: vsock-bridge/src/socket.cpp [hunk 3]
```diff
@@ -17,142 +17,86 @@ namespace vsockio
...
+            Logger::instance->Log(Logger::WARNING, "[socket] error on read, closing (fd=", _fd, "): ", err, ", ", strerror(err));
+            close();
+            return false;
+        }
 	}
 
-	void Socket::send(Buffer& buffer)
+	bool Socket::send(Buffer& buffer)
 	{
+        bool canSendModeData = false;
```
Comment (scong-ttd) [inline vsock-bridge/src/socket.cpp]: typo: "Mode" -> "More".

But reading the method twice, there's only one case we need `canSendModeData` here: when buffer is already consumed entering into the method, where we would return `false`. But this seems to be an exceptional case.

Should we remove this variable and return true at the end of the method?

File: vsock-bridge/src/channel.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#include <channel.h>
+
+namespace vsockio
+{
+    void DirectChannel::performIO()
+    {
+        _a->readInput();
```
Comment (scong-ttd) [inline vsock-bridge/src/channel.cpp:10]: Trying all 4 operations when a channel is ready does add more system calls. I like this tradeoff for simplicity.

**Review by scong-ttd — APPROVED** [review]

**Comment by scong-ttd** [comment]
> - Increased IO buffer size to 1024 bytes

Shall we increase to more based on uid operator payload size? (or make it configurable)

File: vsock-bridge/include/buffer.h [hunk 5]
```diff
@@ -1,222 +1,63 @@
...
-
-		Buffer(Buffer&& b) : _arena(b._arena), _pageCount(b._pageCount), _cursor(b._cursor), _size(b._size), _pageSize(b._arena->blockSize())
-		{
-			for (int i = 0; i < _pageCount; i++)
-			{
-				_pages[i] = b._pages[i];
-			}
-			b._pageCount = 0; // prevent _pages being destructed by old object
-		}
+        static constexpr int BUFFER_SIZE = 4096;
```
Comment (scong-ttd) [inline vsock-bridge/include/buffer.h:12]: In the description you mentioned buffer size 1024? Would it be useful to configure the buffer size for UID2 use cases?

File: vsock-bridge/src/config.cpp (hunk 1 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/config.cpp]: Rewrote function in terms of stoul.

File: vsock-bridge/src/socket.cpp (hunk 2 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp:56]: Fixed tabs to spaces in whole file. Should add clang-style in a future PR.

File: vsock-bridge/src/socket.cpp (hunk 3 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp]: Removed the variable.

File: vsock-bridge/src/channel.cpp (hunk 4 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/channel.cpp:10]: Yep, added a comment to make this a bit clearer.

**Comment by atarassov-ttd** [comment]
> > * Increased IO buffer size to 1024 bytes
> 
> Shall we increase to more based on uid operator payload size? (or make it configurable)

That was incorrect statement in the PR description, updated to 4096 which is the final number we settled on based on perf testing.

File: vsock-bridge/include/buffer.h (hunk 5 above)
Comment (atarassov-ttd) [reply vsock-bridge/include/buffer.h:12]: 4096 is correct. This is the number we settled on based on perf testing. It is equal to the minimum socket send buffer size on Linux (added comment about that). This gives priority and performance for small requests/responses (e.g. token/generate or token/refresh) which tend to be real-time sensitive.

**Review by cody-constine-ttd — APPROVED** [review]

**Pull request merged**